README.rst
.. image:: https://www.repostatus.org/badges/latest/unsupported.svg
    :target: https://www.repostatus.org/#unsupported
    :alt: Project Status: Unsupported – The project has reached a stable,
          usable state but the author(s) have ceased all work on it. A new
          maintainer may be desired.

.. image:: https://img.shields.io/pypi/pyversions/binheat.svg
    :target: https://pypi.org/project/binheat/

.. image:: https://img.shields.io/github/license/jwodder/binheat.svg
    :target: https://opensource.org/licenses/MIT
    :alt: MIT License

`GitHub <https://github.com/jwodder/binheat>`_
| `PyPI <https://pypi.org/project/binheat/>`_
| `Issues <https://github.com/jwodder/binheat/issues>`_
| `Changelog <https://github.com/jwodder/binheat/blob/master/CHANGELOG.md>`_

``binheat`` converts a description of a binary relation into a PDF image of the
relation as a binary heat map (a.k.a. matrix display, adjacency matrix,
comparison chart, and probably a bunch of other names as well; see below for an
example).

Each line of the input (except for blank lines and comments, which are ignored)
must be of the form ``x<TAB>y<TAB>z...``, denoting pairs ``(x, y)``, ``(x,
z)``, etc. in the binary relation.

In the output table, the values from the first column of each input line become
the labels of the table's rows, and the values from the second input column
onwards become the labels of the table's columns.  This can be reversed with
the ``--transpose`` option.


Installation
============
``binheat`` requires Python 3.7 or higher.  Just use `pip
<https://pip.pypa.io>`_ for Python 3 (You have pip, right?) to install
``binheat`` and its dependencies::

    python3 -m pip install binheat


Usage
=====

::

    binheat [<OPTIONS>] [<infile> [<outfile>]]

Input is read from ``<infile>`` (defaulting to standard input), and the
resulting PDF is written to ``<outfile>`` (defaulting to ``<infile>`` with its
file extension changed to ``.pdf``, or to standard output if ``<infile>`` is
standard input).


Options
-------

-C FILE, --column-labels FILE
                        Use the lines in ``FILE`` (after discarding blank lines
                        & comments) in the order they appear as column labels
                        (or row labels if ``--transpose`` is in effect).  Any
                        pairs in the input whose second column does not appear
                        in ``FILE`` are discarded.

-F FONT, --font FONT    Typeset text in the given font.  ``FONT`` must be
                        either the name of a builtin PostScript font or the
                        path to a ``.ttf`` file.  By default, text is typeset
                        in Times-Roman.

-f SIZE, --font-size SIZE
                        Set the text size to ``SIZE`` (default 12).

-R FILE, --row-labels FILE
                        Use the lines in ``FILE`` (after discarding blank lines
                        & comments) in the order they appear as row labels (or
                        column labels if ``--transpose`` is in effect).  Any
                        pairs in the input whose first column does not appear
                        in ``FILE`` are discarded.

--sort, --no-sort       Whether to list labels in the output in lexical order
                        rather than in the order in which they appear in the
                        input file; default: ``--no-sort``

-T, --transpose         The output will be transposed — i.e., the first column
                        of the input will be used for the output table's column
                        labels, and the second input column onwards will be
                        used for the table's row labels.


Example
=======

The following input file::

    NUL (\0, 0x00)<TAB>iscntrl
    0x01..0x06<TAB>iscntrl
    BEL (\a, 0x07)<TAB>iscntrl
    BS (\b, 0x08)<TAB>iscntrl
    TAB (\t, 0x09)<TAB>iscntrl<TAB>isspace<TAB>isblank
    LF (\n, 0x0A)<TAB>iscntrl<TAB>isspace
    VT (\v, 0x0B)<TAB>iscntrl<TAB>isspace
    FF (\f, 0x0C)<TAB>iscntrl<TAB>isspace
    CR (\r, 0x0D)<TAB>iscntrl<TAB>isspace
    0x0E..0x1F<TAB>iscntrl
    SPACE (0x20)<TAB>isprint<TAB>isspace<TAB>isblank
    !"#$%&'()*+,-./<TAB>isprint<TAB>isgraph<TAB>ispunct
    0123456789<TAB>isprint<TAB>isgraph<TAB>isalnum<TAB>isdigit<TAB>isxdigit
    :;<=>?@<TAB>isprint<TAB>isgraph<TAB>ispunct
    ABCDEF<TAB>isprint<TAB>isgraph<TAB>isalnum<TAB>isalpha<TAB>isupper<TAB>isxdigit
    GHIJKLMNOPQRSTUVWXYZ<TAB>isprint<TAB>isgraph<TAB>isalnum<TAB>isalpha<TAB>isupper
    [\]^_`<TAB>isprint<TAB>isgraph<TAB>ispunct
    abcdef<TAB>isprint<TAB>isgraph<TAB>isalnum<TAB>isalpha<TAB>islower<TAB>isxdigit
    ghijklmnopqrstuvwxyz<TAB>isprint<TAB>isgraph<TAB>isalnum<TAB>isalpha<TAB>islower
    {|}~<TAB>isprint<TAB>isgraph<TAB>ispunct
    DEL (0x7F)<TAB>iscntrl

produces (using the default options) an output file that looks like this:

.. image:: https://github.com/jwodder/binheat/raw/master/examples/ctype.png


CHANGELOG.md
v0.2.2 (2023-12-12)
-------------------
- Support Python 3.10, 3.11, and 3.12
- Drop support for Python 3.6
- Migrated from setuptools to hatch
- Declare the project unsupported going forwards

v0.2.1 (2021-05-31)
-------------------
- Drop support for Python 3.4 and 3.5
- Support Python 3.9
- Support Click 8

v0.2.0 (2020-01-24)
-------------------
- `--font` now accepts names of builtin fonts
- When an input filename is given but no output filename, write to the input
  filename with the extension set to `.pdf` instead of to stdout
- Eliminated the `--multiline` option and made its behavior the default
- Added a `--sort` option and made `--no-sort` the default behavior
- Support Python 3.8

v0.1.0 (2019-04-15)
-------------------
Initial release


binheat.pl
#!/usr/bin/perl -w
use strict;
use Getopt::Std;

sub maximum(@);
sub readList($\%);
sub pushPair($$);
sub psify($);

my $font = 'Times-Roman';
my $fontsize = 12;
my $em = 0.7;  # expected em-to-font-size ratio
my $padding = 5;
my $linepad = 2.5;

my(%opts, %lefts, %tops, %matrix);
getopts('1:2:TSm', \%opts) || exit 2;
@opts{'1','2'} = @opts{'2','1'} if $opts{T};

my($rows, $cols) = (0, 0);
$rows = readList $opts{1}, %lefts if defined $opts{1};
$cols = readList $opts{2}, %tops  if defined $opts{2};

my($in, $out) = (*STDIN, *STDOUT);
my $infile = shift || '-';
my $outfile = shift || '-';
open $in,  '<', $infile  or die "$0: $infile: $!"  unless $infile  eq '-';
open $out, '>', $outfile or die "$0: $outfile: $!" unless $outfile eq '-';
select $out;

while (<$in>) {
 chomp;
 s/\s+$//;
 next if /^$/ || /^\s*#/;
 my($left, @top) = split /\t+/;
 if ($opts{m}) { pushPair($left, $_) for @top }
 elsif (@top) { pushPair($left, $top[0]) }
}

if (!$opts{S}) {
 if (!defined $opts{1}) {
  my $i = 0;
  %lefts = map { $_ => $i++ } sort keys %matrix;
  $rows = $i;
 }
 if (!defined $opts{2}) {
  my $i = 0;
  %tops = map { %$_ } values %matrix;
  $tops{$_} = $i++ for sort keys %tops;
  $cols = $i;
 }
}

### TODO: These guesses should use the lengths of the Latin-1 encoded labels.
my $leftlen = maximum(map length, keys %lefts) * $em * $fontsize;
my $toplen  = maximum(map length, keys %tops)  * $em * $fontsize;

my $miny = $rows * $fontsize * 1.2;
my $maxx = $cols * $fontsize * 1.2;

print <<EOT;
%!PS-Adobe-3.0 EPSF-3.0
%%BoundingBox: -@{[$leftlen+$padding]} -@{[$miny+$padding]} @{[$maxx+$padding]} @{[$toplen+$padding]}

/mkLatin1 {  % old font, new name -- new font
 exch dup length dict begin
 { 1 index /FID ne { def } { pop pop } ifelse } forall
 /Encoding ISOLatin1Encoding def
 currentdict end definefont
} def

/fontsize $fontsize def
/$font findfont /binheat-font mkLatin1 fontsize scalefont setfont
/lineheight fontsize 1.2 mul def
/radius lineheight 3 div def

0.8 0.8 0.8 setrgbcolor
0 2 $cols 1 sub {
 lineheight mul -$miny lineheight $miny $toplen add rectfill
} for

1 1 0.5 setrgbcolor
2 2 $rows {
 % Yes, it starts at 2, so that the positive rectfill height will make it fill
 % row 1.
 lineheight mul neg -$leftlen exch $leftlen $maxx add lineheight rectfill
} for

0 0 0 setrgbcolor
newpath 0   $toplen moveto 0 -$miny lineto stroke
newpath -$leftlen 0 moveto $maxx  0 lineto stroke

/dot {
 0.5 add lineheight mul exch 0.5 add lineheight mul neg radius 0 360 arc fill
} def
EOT

while (my($key, $i) = each %lefts) {
 my $left = psify $key;
 print <<EOT;
($left) dup stringwidth pop neg $linepad sub
$i 1 add lineheight mul neg fontsize 3 div add
moveto
show
EOT
}

while (my($key, $i) = each %tops) {
 my $top = psify $key;
 print <<EOT;
gsave
$i 1 add lineheight mul 0 translate 90 rotate $linepad fontsize 3 div moveto
($top) show
grestore
EOT
}

for (grep { exists $lefts{$_} } keys %matrix) {
 my $rowNo = $lefts{$_};
 print "$rowNo $tops{$_} dot\n"
  for grep { exists $tops{$_} } keys %{$matrix{$_}};
}

print "showpage\n";

sub maximum(@) {
 my $max = shift;
 for (@_) { $max = $_ if $_ > $max }
 return $max;
}

sub readList($\%) {
 my($file, $hash) = @_;
 my $fp;
 if ($file eq '-') { $fp = *STDIN }
 else { open $fp, '<', $file or die "$0: $file: $!" }
 my $i = 0;
 while (<$fp>) {
  chomp;
  s/\s+$//;
  next if /^$/ || /^\s*#/;
  $hash->{$_} = $i++;
 }
 return $i;
}

sub pushPair($$) {
 my($left, $top) = @_;
 ($left, $top) = ($top, $left) if $opts{T};
 $matrix{$left}{$top} = 1;
 $lefts{$left} = $rows++
  if $opts{S} && !defined $opts{1} && !exists $lefts{$left};
 $tops{$top} = $cols++ if $opts{S} && !defined $opts{2} && !exists $tops{$top};
}

sub psify($) {
 my $str = shift;
 $str =~ s/([(\\)])/\\$1/g;
 $str =~ s/([\xC0-\xDF])([\x80-\xBF])/
  my $x = (ord($1) & 0x1F) << 6 | (ord($2) & 0x3F);
  $x < 256 ? sprintf("\\%03o", $x) : sprintf("\\%03o\\%03o", ord $1, ord $2);
 /ge;
 $str =~ s/([^\x00-\x7E])/sprintf "\\%03o", ord $1/ge;
 return $str;
}

__END__

=pod

=head1 NAME

B<binheat> - binary heat map generator

=head1 SYNOPSIS

B<binheat> [B<-1> I<file>] [B<-2> I<file>] [B<-mST>] [I<infile> [I<outfile>]]

=head1 DESCRIPTION

B<binheat> converts a description of a binary relation into an Embedded
PostScript image of the relation as a binary heat map (a.k.a. matrix display,
adjacency matrix, comparison chart, and probably a bunch of other names as
well).  I<infile> (or standard input if no file is specified) must list the
elements of the relation, one per line, each line consisting of two labels
(nonempty nontab character sequences) separated by one or more tabs.  (The
input may also contain comment lines in which the first nonwhitespace character
is C<#>.)  B<binheat> will then print an EPS file to I<outfile> (or standard
output if no file is specified) showing a table in which the labels from the
first column of each line are lexically sorted along the left edge, the labels
from the second column are lexically sorted along the top edge, and dots are
placed in the cells of the table to represent the elements of the relation.

=head1 OPTIONS

=over

=item B<-1> I<file>

The lines of I<file> will be taken as a list of all labels appearing in the
first column of the input file, and the labels along the left edge of the
output chart (or the top edge if B<-T> is in effect) will be in the same order
that they are listed in I<file>.  Labels that appear in I<file> but not the
first column of the input file will appear in the output with no relations, and
labels that appear in the first column of the input file but not I<file> will
not appear in the output at all.

This option overrides the B<-S> option for the first column only.

=item B<-2> I<file>

The lines of I<file> will be taken as a list of all labels appearing in the
second column of the input file, and the labels along the top edge of the
output chart (or the left edge if B<-T> is in effect) will be in the same order
that they are listed in I<file>.  Labels that appear in I<file> but not the
second column of the input file will appear in the output with no relations,
and labels that appear in the second column of the input file but not I<file>
will not appear in the output at all.

This option overrides the B<-S> option for the second column only.

=item B<-m>

C<< foo<TAB>bar<TAB>baz >> (or any number of tab-separated fields) will be
allowed as an abbreviation for C<< foo<TAB>bar >> followed by C<< foo<TAB>baz
>> etc.

=item B<-S>

Labels in the output will be listed in the order in which they appear in the
input file rather than in lexical order

=item B<-T>

The output will be transposed -- i.e., the first column will be used for the
top edge of the chart and the second column for the left edge

=back

=head1 RESTRICTIONS

B<binheat> only accepts input in the UTF-8 and ISO-8859-1 ("Latin-1")
encodings, and Latin-1 that looks like UTF-8 (even in part) will confuse it.
More importantly, B<binheat> can only display characters in the Latin-1 range;
anything beyond that will be represented by the Latin-1 interpretation of the
character's UTF-8 encoding.  (For example, a "é" in the input will remain a "é"
in the output, but "α" will become "Î±".)

=head1 AUTHOR

John T. Wodder II <jwodder@sdf.lonestar.org>

=cut


dotmatrix.pl
#!/usr/bin/perl -w
# Options:
#   -1 file - list of all values for the first column in display order
#   -2 file - list of all values for the second column in display order
#   -T - use the first column for the top edge of the chart and the second
#        column for the left edge rather than the other way around
#   -S - display values in the order in which they appear in the input file
#        rather than sorted lexically (overridden by -1 and -2 options)
#   -m - allow "foo\tbar\tbaz\tglarch\n" as shorthand for
#        "\foo\tbar\nfoo\tbaz\nfoo\tglarch\n"
use strict;
use Getopt::Std;

sub maximum(@);
sub readList($\%);
sub pushPair($$);

my $font = 'Times-Roman';
my $fontsize = 12;
my $em = 0.6;  # expected em-to-font-size ratio
my $padding = 5;
my $linepad = 2.5;

my(%opts, %lefts, %tops, %matrix);
getopts('1:2:TSm', \%opts) || exit 2;
@opts{'1','2'} = @opts{'2','1'} if $opts{T};

my($rows, $cols) = (0, 0);
$rows = readList $opts{1}, %lefts if defined $opts{1};
$cols = readList $opts{2}, %tops  if defined $opts{2};

while (<>) {
 chomp;
 s/\s+$//;
 next if /^$/ || /^\s*#/;
 my($left, @top) = split /\t+/;
 if ($opts{m}) { pushPair($left, $_) for @top }
 elsif (@top) { pushPair($left, $top[0]) }
}

if (!$opts{S}) {
 if (!defined $opts{1}) {
  my $i = 0;
  %lefts = map { $_ => $i++ } sort keys %matrix;
  $rows = $i;
 }
 if (!defined $opts{2}) {
  my $i = 0;
  %tops = map { %$_ } values %matrix;
  $tops{$_} = $i++ for sort keys %tops;
  $cols = $i;
 }
}

my $leftlen = maximum(map length, keys %lefts) * $em * $fontsize;
my $toplen  = maximum(map length, keys %tops)  * $em * $fontsize;

my $miny = $rows * $fontsize * 1.2;
my $maxx = $cols * $fontsize * 1.2;

print <<EOT;
%!PS-Adobe-3.0 EPSF-3.0
%%BoundingBox: -@{[$leftlen+$padding]} -@{[$miny+$padding]} @{[$maxx+$padding]} @{[$toplen+$padding]}
/fontsize $fontsize def
/$font findfont fontsize scalefont setfont
/lineheight fontsize 1.2 mul def
/radius lineheight 3 div def

0.8 0.8 0.8 setrgbcolor
0 2 $cols 1 sub {
 lineheight mul -$miny lineheight $miny $toplen add rectfill
} for

1 1 0.5 setrgbcolor
2 2 $rows {
 % Yes, it starts at 2, so that the positive rectfill height will make it fill
 % row 1.
 lineheight mul neg -$leftlen exch $leftlen $maxx add lineheight rectfill
} for

0 0 0 setrgbcolor
newpath 0   $toplen moveto 0 -$miny lineto stroke
newpath -$leftlen 0 moveto $maxx  0 lineto stroke

/dot {
 0.5 add lineheight mul exch 0.5 add lineheight mul neg radius 0 360 arc fill
} def
EOT

while (my($key, $i) = each %lefts) {
 (my $left = $key) =~ s/([(\\)])/\\$1/g;
 print <<EOT;
($left) dup stringwidth pop neg $linepad sub
$i 1 add lineheight mul neg fontsize 3 div add
moveto
show
EOT
}

while (my($key, $i) = each %tops) {
 (my $top = $key) =~ s/([(\\)])/\\$1/g;
 print <<EOT;
gsave
$i 1 add lineheight mul 0 translate 90 rotate $linepad fontsize 3 div moveto
($top) show
grestore
EOT
}

for (grep { exists $lefts{$_} } keys %matrix) {
 my $rowNo = $lefts{$_};
 print "$rowNo $tops{$_} dot\n"
  for grep { exists $tops{$_} } keys %{$matrix{$_}};
}

print "showpage\n";

sub maximum(@) {
 my $max = shift;
 for (@_) { $max = $_ if $_ > $max }
 return $max;
}

sub readList($\%) {
 my($file, $hash) = @_;
 my $in;
 if ($file eq '-') { $in = *STDIN }
 else { open $in, '<', $file or die "$0: $file: $!" }
 my $i = 0;
 while (<$in>) {
  chomp;
  s/\s+$//;
  next if /^$/ || /^\s*#/;
  $hash->{$_} = $i++;
 }
 return $i;
}

sub pushPair($$) {
 my($left, $top) = @_;
 ($left, $top) = ($top, $left) if $opts{T};
 $matrix{$left}{$top} = 1;
 $lefts{$left} = $rows++
  if $opts{S} && !defined $opts{1} && !exists $lefts{$left};
 $tops{$top} = $cols++ if $opts{S} && !defined $opts{2} && !exists $tops{$top};
}


binheat.py
#!/usr/bin/python3
"""
Binary heat map generator

``binheat`` converts a description of a binary relation into a PDF image of the
relation as a binary heat map (a.k.a. matrix display, adjacency matrix,
comparison chart, and probably a bunch of other names as well).

Each line of the input (except for blank lines and comments, which are ignored)
must be of the form ``x<TAB>y<TAB>z...``, denoting pairs ``(x, y)``, ``(x,
z)``, etc. in the binary relation.

In the output table, the values from the first column of each input line become
the labels of the table's rows, and the values from the second input column
onwards become the labels of the table's columns.  This can be reversed with
the ``--transpose`` option.

Run ``binheat --help`` or visit <https://github.com/jwodder/binheat> for more
information.
"""

__version__ = "0.2.2"
__author__ = "John Thorvald Wodder II"
__author_email__ = "binheat@varonathe.org"
__license__ = "MIT"
__url__ = "https://github.com/jwodder/binheat"

from pathlib import Path
import re
import sys
import click
from reportlab.pdfbase import pdfmetrics
from reportlab.pdfbase.ttfonts import TTFont
from reportlab.pdfgen.canvas import Canvas

#: Padding (in points) around the edge of the rendered binary heat map
PADDING = 5

#: Padding (in points) at the beginning & ending of the row & column labels
LABEL_PAD = 2.5

#: Color to use to highlight alternating columns
COL_BG_COLOR = (0.8, 0.8, 0.8)  # grey

#: Color to use to highlight alternating rows
ROW_BG_COLOR = (1, 1, 0.5)  # light yellow


class BinHeat:
    def __init__(self, allow_extra=False):
        #: Whether to allow (True) or discard (False) pairs containing an
        #: unknown label.  Only has an effect when
        #: `set_row_labels()`/`set_column_labels()` has been called.
        self.allow_extra = allow_extra
        #: Whether `set_row_labels()` has been called
        self.row_labels_set = False
        #: Whether `set_column_labels()` has been called
        self.column_labels_set = False
        #: Mapping from row labels to their indices when listed
        self.rows2indices = {}
        #: Mapping from column labels to their indices when listed
        self.columns2indices = {}
        #: Set of ``(row label, column label)`` pairs
        self.pairs = set()

    def set_row_labels(self, labels):
        """
        Set the row labels to the given sequence of strings in the order given

        :param labels: an iterable of strings to set the row labels to
        :raises RuntimeError: if called more than once or after calling
            `add_pair()`
        """
        if self.row_labels_set:
            raise RuntimeError("set_row_labels() called more than once")
        elif self.pairs:
            raise RuntimeError("set_row_labels() called after add_pair()")
        else:
            self.row_labels_set = True
        self.rows2indices = {l: i for i, l in enumerate(labels)}

    def set_column_labels(self, labels):
        """
        Set the column labels to the given sequence of strings in the order
        given

        :param labels: an iterable of strings to set the column labels to
        :raises RuntimeError: if called more than once or after calling
            `add_pair()`
        """
        if self.column_labels_set:
            raise RuntimeError("set_column_labels() called more than once")
        elif self.pairs:
            raise RuntimeError("set_column_labels() called after add_pair()")
        else:
            self.column_labels_set = True
        self.columns2indices = {l: i for i, l in enumerate(labels)}

    def add_pair(self, row, column):
        """
        Register a pairing of the given row and column.  If `allow_extra` is
        false and either component does not appear in a prespecified label set,
        the pair is discarded.  If `allow_extra` is true, new row or column
        labels are appended to the list of all row/column labels.

        :param str row: the label of the row of this point
        :param str column: the label of the column of this point
        """
        if not self.allow_extra:
            if self.row_labels_set and row not in self.rows2indices:
                return
            if self.column_labels_set and column not in self.columns2indices:
                return
        if row not in self.rows2indices:
            self.rows2indices[row] = len(self.rows2indices)
        if column not in self.columns2indices:
            self.columns2indices[column] = len(self.columns2indices)
        self.pairs.add((row, column))

    @property
    def rows(self):
        """The number of distinct row labels in the binary heat map"""
        return len(self.rows2indices)

    @property
    def columns(self):
        """The number of distinct column labels in the binary heat map"""
        return len(self.columns2indices)

    @property
    def row_labels(self):
        """A list of all row labels in the binary heat map in display order"""
        return sorted(
            self.rows2indices.keys(),
            key=self.rows2indices.__getitem__,
        )

    @property
    def column_labels(self):
        """
        A list of all column labels in the binary heat map in display order
        """
        return sorted(
            self.columns2indices.keys(),
            key=self.columns2indices.__getitem__,
        )

    def get_indexed_pairs(self):
        """
        Yield all pairs in the binary heat map in unspecified order, with each
        pair represented as a pairing of its row's index and its column's index
        """
        for r, c in self.pairs:
            yield (self.rows2indices[r], self.columns2indices[c])

    def sort_labels(self):
        """
        If `set_row_labels()` has not been called, reorder (and reindex) the
        row labels to be in lexicographic order.  Likewise for
        `set_column_labels()` and the column labels.
        """
        ### TODO: Take allow_extra into account when *_labels_set
        if not self.row_labels_set:
            self.rows2indices = {
                k: i for i, k in enumerate(sorted(self.rows2indices.keys()))
            }
        if not self.column_labels_set:
            self.columns2indices = {
                k: i for i, k in enumerate(sorted(self.columns2indices.keys()))
            }

    def render(self, outfile, font_name, font_size):
        """
        Render the binary heat map as a PDF to the file-like object or filename
        ``outfile``.  All text will be typeset in the font named ``font_name``
        at size ``font_size``.
        """
        c = Canvas(outfile)
        c.setFont(font_name, font_size)
        leftlen = max(map(c.stringWidth, self.row_labels)) + LABEL_PAD * 2
        toplen = max(map(c.stringWidth, self.column_labels)) + LABEL_PAD * 2
        miny = self.rows * font_size * 1.2
        maxx = self.columns * font_size * 1.2
        c.setPageSize((leftlen + maxx + PADDING * 2, miny + toplen + PADDING * 2))
        # Set coordinates so that LL corner has coord (-leftlen-PADDING,
        # -miny-PADDING) and the origin is at the point where the borders of the
        # row & column labels meet:
        c.translate(leftlen + PADDING, miny + PADDING)

        lineheight = font_size * 1.2
        radius = lineheight / 3

        c.setFillColorRGB(*COL_BG_COLOR)
        for i in range(0, self.columns, 2):
            c.rect(
                i * lineheight,
                -miny,
                lineheight,
                miny + toplen,
                stroke=0,
                fill=1,
            )

        c.setFillColorRGB(*ROW_BG_COLOR)
        for i in range(2, self.rows + 1, 2):
            # Yes, it starts at 2, so that the positive rectangle height will
            # make it fill row 1.
            c.rect(
                -leftlen,
                -i * lineheight,
                leftlen + maxx,
                lineheight,
                stroke=0,
                fill=1,
            )

        c.setFillColorRGB(0, 0, 0)
        c.line(0, toplen, 0, -miny)
        c.line(-leftlen, 0, maxx, 0)

        for i, label in enumerate(self.row_labels):
            c.drawRightString(
                -LABEL_PAD,
                -(i + 1) * lineheight + font_size / 3,
                label,
            )

        for i, label in enumerate(self.column_labels):
            c.saveState()
            c.translate((i + 1) * lineheight, 0)
            c.rotate(90)
            c.drawString(LABEL_PAD, font_size / 3, label)
            c.restoreState()

        for row, col in self.get_indexed_pairs():
            c.circle(
                (col + 0.5) * lineheight,
                -(row + 0.5) * lineheight,
                radius,
                stroke=0,
                fill=1,
            )

        c.showPage()
        c.save()


@click.command(context_settings={"help_option_names": ["-h", "--help"]})
@click.option(
    "-C",
    "--column-labels",
    type=click.File(),
    help="Use lines in given file as column labels",
)
@click.option(
    "-F",
    "--font",
    default="Times-Roman",
    show_default=True,
    help="Typeset text in given font",
    metavar="NAME|TTF_FILE",
)
@click.option(
    "-f",
    "--font-size",
    type=float,
    default=12,
    show_default=True,
    help="Typeset text at given size",
)
@click.option(
    "-R",
    "--row-labels",
    type=click.File(),
    help="Use lines in given file as row labels",
)
@click.option(
    "--sort/--no-sort",
    "-s/-S",
    "to_sort",
    is_flag=True,
    help="Sort the row & column labels",
)
@click.option("-T", "--transpose", is_flag=True, help="Exchange rows with columns")
@click.version_option(__version__, "-V", "--version", message="binheat %(version)s")
@click.argument("infile", type=click.File(), default="-")
@click.argument("outfile", type=click.File("wb"), required=False)
def main(
    infile, outfile, font, font_size, transpose, to_sort, row_labels, column_labels
):
    """
    Binary heat map generator

    ``binheat`` converts a description of a binary relation into a PDF image of
    the relation as a binary heat map (a.k.a. matrix display, adjacency matrix,
    comparison chart, and probably a bunch of other names as well).

    Each line of the input (except for blank lines and comments, which are
    ignored) must be of the form ``x<TAB>y<TAB>z...``, denoting pairs ``(x,
    y)``, ``(x, z)``, etc. in the binary relation.

    In the output table, the values from the first column of each input line
    become the labels of the table's rows, and the values from the second input
    column onwards become the labels of the table's columns.  This can be
    reversed with the ``--transpose`` option.

    Visit <https://github.com/jwodder/binheat> for more information.
    """
    if font in available_fonts():
        font_name = font
    else:
        # Assume we've been given a path to a .ttf file
        font_name = "CustomFont"
        ### TODO: Use the basename of the filename as the font name?  (Could
        ### that ever cause problems?)
        pdfmetrics.registerFont(TTFont(font_name, font))

    bh = BinHeat()

    if transpose:
        row_labels, column_labels = column_labels, row_labels
    if row_labels:
        bh.set_row_labels(strip_read(row_labels))
    if column_labels:
        bh.set_column_labels(strip_read(column_labels))

    for line in strip_read(infile):
        left, *top = re.split(r"\t+", line)
        for t in top:
            if transpose:
                bh.add_pair(t, left)
            else:
                bh.add_pair(left, t)

    if to_sort:
        bh.sort_labels()

    if outfile is None:
        if infile is sys.stdin:
            outfile_name = "-"
        else:
            outfile_name = str(Path(infile.name).with_suffix(".pdf"))
        outfile = click.open_file(outfile_name, "wb")

    bh.render(outfile, font_name, font_size)


def strip_read(fp):
    r"""
    Yield the lines in file ``fp`` with trailing whitespace stripped and with
    lines beginning with ``/^\s*#/`` (i.e., comment lines) discarded
    """
    for line in fp:
        line = line.rstrip()
        if line == "" or line.lstrip().startswith("#"):
            continue
        yield line


def available_fonts():
    return Canvas("").getAvailableFonts()
    # return pdfmetrics.standardFonts


if __name__ == "__main__":
    main()
